gossip/src/duplicate_shred_handler.rs

```diff
@@ -0,0 +1,431 @@
+use {
+    crate::{
+        duplicate_shred::{into_shreds, DuplicateShred, Error},
+        duplicate_shred_listener::DuplicateShredHandlerTrait,
+    },
+    log::*,
+    solana_ledger::{blockstore::Blockstore, leader_schedule_cache::LeaderScheduleCache},
+    solana_sdk::{clock::Slot, pubkey::Pubkey},
+    std::{
+        collections::{HashMap, HashSet},
+        sync::Arc,
+    },
+};
+
+const CLEANUP_EVERY_N_LOOPS: usize = 10;
+// Normally num_chunks is 3, because there are two shreds (each is one packet)
+// and meta data. So we discard anything larger than 3 chunks.
+const MAX_NUM_CHUNKS: u8 = 3;
+// We only allow each pubkey to send proofs for 5 slots, because normally there
+// is only 1 person sending out duplicate proofs, 1 person is leader for 4 slots,
+// so we allow 5 here to limit the chunk map size.
+const ALLOWED_SLOTS_PER_PUBKEY: usize = 5;
+
+struct ProofChunkMap {
+    num_chunks: u8,
+    missing_chunks: u8,
+    wallclock: u64,
+    chunks: HashMap<u8, DuplicateShred>,
+}
+
+// Group received chunks by peer pubkey, when we receive an invalid proof,
+// set the value to Frozen so we don't accept future proofs with the same key.
+type SlotChunkMap = HashMap<Pubkey, ProofChunkMap>;
+
+enum SlotStatus {
+    Frozen,
+    UnfinishedProof(SlotChunkMap),
+}
+pub struct DuplicateShredHandler {
+    // Because we use UDP for packet transfer, we can normally only send ~1500 bytes
+    // in each packet. We send both shreds and meta data in duplicate shred proof, and
+    // each shred is normally 1 packet(1500 bytes), so the whole proof is larger than
+    // 1 packet and it needs to be cut down as chunks for transfer. So we need to piece
+    // together the chunks into the original proof before anything useful is done.
+    //
+    // When a valid proof has been inserted, we change the entry for that slot to Frozen
+    // to indicate we no longer accept proofs for this slot.
+    chunk_map: HashMap<Slot, SlotStatus>,
+    // We don't want bad guys to inflate the chunk map, so we limit the number of
+    // pending proofs from each pubkey to ALLOWED_SLOTS_PER_PUBKEY.
+    validator_pending_proof_map: HashMap<Pubkey, HashSet<Slot>>,
```

Comment:
I don't have a good answer here, because these pubkeys come from Gossip, so whatever I can verify will not be more advanced than Gossip. The only thing I can do is to limit number of PubKeys I accept here, but then a bad guy with many PubKeys can easily flush the map with garbage. What would you suggest here?

Reply:
maybe limit to pubkeys that are part of leader schedule / have voted recently etc.

Reply:
Yeah I think  we unfortunately have to keep proofs for anything within an epoch of the root (essentially slots that we know the leader schedule for).

Instead of limits like `ALLOWED_SLOTS_PER_PUBKEY` and `MAX_SLOT_DISTANCE_TO_ROOT` on the stored proofs it might be better to offload this storage to disk through extending the column family in Rocks

This can be handled in another PR though